File: src/ell/providers/openai.py
from abc import ABC, abstractmethod
from collections import defaultdict
from typing import Any, Callable, Dict, List, Optional, Tuple, Type, Union, cast

from pydantic import BaseModel
from ell.provider import  EllCallParams, Metadata, Provider
from ell.types import Message, ContentBlock, ToolCall
from ell.types._lstr import _lstr
import json
from ell.configurator import _Model, config, register_provider
from ell.types.message import LMP
from ell.util.serialization import serialize_image

try: 
    # XXX: Could genericize.
    import openai
    from openai._streaming import Stream
    from openai.types.chat import ChatCompletion, ParsedChatCompletion, ChatCompletionChunk, ChatCompletionMessageParam

    class OpenAIProvider(Provider):
        dangerous_disable_validation = True
        
        def provider_call_function(self, client : openai.Client, api_call_params : Optional[Dict[str, Any]] = None) -> Callable[..., Any]:
            if api_call_params and (isinstance(fmt := api_call_params.get("response_format"), type)) and issubclass(fmt, BaseModel):
                return client.beta.chat.completions.parse
            else:
                return client.chat.completions.create
            
        def translate_to_provider(self, ell_call : EllCallParams) -> Dict[str, Any]: 
            final_call_params = ell_call.api_params.copy()
            final_call_params["model"] = ell_call.model
            # Stream by default for verbose logging.
            final_call_params["stream"] = True
            final_call_params["stream_options"] = {"include_usage": True}

            # XXX: Deprecation of config.registry.supports_streaming when streaming is implemented.
            if ell_call.tools or final_call_params.get("response_format") or (regisered_model := config.registry.get(ell_call.model, None)) and regisered_model.supports_streaming is False:
                final_call_params.pop("stream", None)
                final_call_params.pop("stream_options", None)
            if ell_call.tools:
                final_call_params.update(
                    tool_choice="auto",
                    tools=[  
                        dict(
                            type="function",
                            function=dict(
                                name=tool.__name__,
                                description=tool.__doc__,
                                parameters=tool.__ell_params_model__.model_json_schema(),  #type: ignore
                            )
                        ) for tool in ell_call.tools
                    ]
                )
            # messages
            openai_messages : List[ChatCompletionMessageParam] = []
            for message in ell_call.messages:
                if (tool_calls := message.tool_calls):
                    assert message.role == "assistant", "Tool calls must be from the assistant."
                    assert all(t.tool_call_id for t in tool_calls), "Tool calls must have tool call ids."
                    openai_messages.append(dict(
                        tool_calls=[
                            dict(
                                id=cast(str, tool_call.tool_call_id),
                                type="function",
                                function=dict(
                                    name=tool_call.tool.__name__,
                                    arguments=json.dumps(tool_call.params.model_dump())
                                )
                            ) for tool_call in tool_calls ],
                        role="assistant",
                        content=None,
                    ))
                elif (tool_results := message.tool_results):
                    for tool_result in tool_results:
                        assert all(cb.type == "text" for cb in tool_result.result), "Tool result does not match expected content blocks."
                        openai_messages.append(dict(
                            role="tool",
                            tool_call_id=tool_result.tool_call_id,
                            content=tool_result.text_only, 
                        ))
                else:
                    openai_messages.append(cast(ChatCompletionMessageParam, dict(
                        role=message.role,
                        content=[_content_block_to_openai_format(c) for c in message.content] 
                             if message.role != "system" 
                             else message.text_only
                    )))
                     
            final_call_params["messages"] = openai_messages
            
            return final_call_params
        
        def translate_from_provider(
            self,
            provider_response: Union[
                ChatCompletion, 
                ParsedChatCompletion,
                Stream[ChatCompletionChunk], Any],
            ell_call: EllCallParams,
            provider_call_params: Dict[str, Any],
            origin_id: Optional[str] = None,
            logger: Optional[Callable[..., None]] = None,
        ) -> Tuple[List[Message], Metadata]:
            
            metadata : Metadata = {} 
            messages : List[Message] = []
            did_stream = provider_call_params.get("stream", False)

        
            if did_stream:
                stream = cast(Stream[ChatCompletionChunk], provider_response)
                message_streams = defaultdict(list)
                role : Optional[str] = None
                for chunk in stream: 
                    metadata.update(chunk.model_dump(exclude={"choices"})) 
                    
                    for chat_compl_chunk in chunk.choices:
                        message_streams[chat_compl_chunk.index].append(chat_compl_chunk)
                        delta = chat_compl_chunk.delta
                        role = role or delta.role
                        if  chat_compl_chunk.index == 0 and logger:
                            logger(delta.content, is_refusal=hasattr(delta, "refusal") and delta.refusal)
                for _, message_stream in sorted(message_streams.items(), key=lambda x: x[0]):
                    text = "".join((choice.delta.content or "") for choice in message_stream)
                    messages.append(
                        Message(role=role, 
                                content=_lstr(content=text,origin_trace=origin_id)))
                    #XXX: Support streaming other types.
            else:
                chat_completion = cast(Union[ChatCompletion, ParsedChatCompletion], provider_response)
                metadata = chat_completion.model_dump(exclude={"choices"})
                for oai_choice in chat_completion.choices: 
                    role = oai_choice.message.role
                    content_blocks = []
                    if (hasattr(message := oai_choice.message, "refusal") and (refusal := message.refusal)):
                        raise ValueError(refusal)
                    if hasattr(message, "parsed"):
                        if (parsed := message.parsed): 
                            content_blocks.append(ContentBlock(parsed=parsed)) #XXX: Origin tracing
                            if logger: logger(parsed.model_dump_json())
                    else:
                        if (content := message.content):
                            content_blocks.append(
                                ContentBlock(
                                    text=_lstr(content=content,origin_trace=origin_id)))
                            if logger: logger(content)
                        if (tool_calls := message.tool_calls):
                            for tool_call in tool_calls:
                                matching_tool = ell_call.get_tool_by_name(tool_call.function.name)
                                assert matching_tool, "Model called tool not found in provided toolset."
                                content_blocks.append(
                                    ContentBlock(
                                        tool_call=ToolCall(
                                            tool=matching_tool,
                                            tool_call_id=_lstr(
                                                tool_call.id, origin_trace= origin_id),
                                            params=json.loads(tool_call.function.arguments),
                                        )
                                    )
                                )
                                if logger: logger(repr(tool_call))
                    messages.append(Message(role=role, content=content_blocks))
            return messages, metadata


    # xx: singleton needed
    openai_provider = OpenAIProvider()
    register_provider(openai_provider, openai.Client)
except ImportError:
    pass

def _content_block_to_openai_format(content_block: ContentBlock) -> Dict[str, Any]:
    if (image := content_block.image):
        image_url = dict(url=serialize_image(image.image) if image.image else image.url)
        # XXX: Solve per content params better
        if image.detail: image_url["detail"] = image.detail
        return {
            "type": "image_url",
            "image_url": image_url
        }
    elif ((text := content_block.text) is not None): return dict(type="text", text=text)
    elif (parsed := content_block.parsed): return dict(type="text", text=parsed.model_dump_json())    
    else:
        raise ValueError(f"Unsupported content block type for openai: {content_block}")

File: typescript/src/providers/openai.ts
try {
  require('openai')
} catch (e) {
  console.error('OpenAI not found. Please install it to use the OpenAI provider.')
}
import { APICallResult, BaseProvider, EllCallParams, Metadata, Provider } from '../provider'
import { Message, ContentBlock, ToolCall } from '../types'
import { LMP } from '../types/message'
// import { serializeImage } from "../util/serialization";
import { config, registerProvider } from '../configurator'
import OpenAI from 'openai'
// TODO>
const serializeImage = (image: any) => {
  return image
  // return `data:image/jpeg;base64,${image.data}`;
}

import { Stream as OpenAIStream } from 'openai/streaming'
import { zodFunction, zodResponseFormat } from 'openai/helpers/zod'
import { Tool } from '../types/tools'
import { ZodAny, ZodType } from 'zod'
import * as logging from '../util/_logging'

const logger = logging.getLogger('openai-provider')

type OpenAIResponseFormat =
  | OpenAI.ResponseFormatText
  | OpenAI.ResponseFormatJSONObject
  | OpenAI.ResponseFormatJSONSchema

class _lstr {
  constructor(
    public value: string,
    public _originTrace: string
  ) {}
}

const mapResponseFormat = (responseFormat: unknown): OpenAIResponseFormat | undefined => {
  if (responseFormat instanceof ZodType) {
    return zodResponseFormat(responseFormat, 'ZodType')
  }

  // @ts-ignore we trust the user here for now
  return responseFormat
}

function omit<T extends object, K extends keyof T>(obj: T, keys: K[]): Omit<T, K> {
  const result = { ...obj }
  for (const key of keys) {
    delete result[key]
  }
  return result
}

const mapToolDefinition = (tool: Tool<any, any>): OpenAI.Chat.Completions.ChatCompletionTool => {
  if (tool.__ell_tool_input__ instanceof ZodType && tool.__ell_tool_output__ instanceof ZodType) {
    return zodFunction({
      name: tool.__ell_tool_name__,
      parameters: tool.__ell_tool_input__,
      description: tool.__ell_tool_description__,
      function: tool,
    })
  }

  throw new Error(`Unsupported tool definition for openai: ${tool}`)
}

const getCanStream = (ellCall: EllCallParams) => {
  if (ellCall.tools && ellCall.tools.length > 0) {
    return false
  }
  if (ellCall.apiParams.response_format) {
    return false
  }
  return true
}
const mapToStreamingParams = async (
  ellCall: EllCallParams
): Promise<OpenAI.Chat.Completions.ChatCompletionCreateParamsStreaming> => {
  const messages = await Promise.all(ellCall.messages.map((message) => messageToOpenAIFormat(message)))

  return {
    model: ellCall.model,
    ...omit(ellCall.apiParams, ['tools', 'response_format', 'messages', 'stream', 'stream_options']),
    messages,
    response_format: mapResponseFormat(ellCall.apiParams.response_format),
    stream: true,
    stream_options: { include_usage: true },
  }
}

const mapToNonStreamingParams = async (
  ellCall: EllCallParams
): Promise<OpenAI.Chat.Completions.ChatCompletionCreateParamsNonStreaming> => {
  const messages = await Promise.all(ellCall.messages.map((message) => messageToOpenAIFormat(message)))
  return {
    model: ellCall.model,
    messages,
    tool_choice: ellCall.tools && ellCall.tools.length > 0 ? 'auto' : undefined,
    tools: ellCall.tools?.map((tool) => mapToolDefinition(tool)),
    response_format: mapResponseFormat(ellCall.apiParams.response_format),
    ...omit(ellCall.apiParams, ['tools', 'response_format', 'messages', 'stream', 'stream_options']),
    stream: false,
  } as OpenAI.Chat.Completions.ChatCompletionCreateParamsNonStreaming
}

class _OpenAIProvider extends BaseProvider implements Provider {
  constructor() {
    super()
  }
  dangerous_disable_validation = true

  // TODO. needed?
  disallowedApiParams() {
    return new Set([])
  }
  // TODO. needed? types?
  availableApiParams(client: any, api_call_params: Record<string, any>): string[] {
    return []
  }

  // TODO. needed? only public api is "call"
  providerCallFunction(client: any, api_call_params: Record<string, any>): (...args: any[]) => Promise<any> {
    if (api_call_params.response_format) {
      return (...args: any[]) => client.beta.chat.completions.parse(...args)
    }
    return (...args: any[]) => client.chat.completions.create(...args)
  }

  async call(
    params: EllCallParams,
    origin_id: string,
    logger?: any
  ): Promise<[Message[], Record<string, any>, Metadata]> {
    const input = await this.translateToProvider(params)

    const callAPI = this.providerCallFunction(params.client, input)

    const output = await callAPI(input)

    const [messages, metadata] = await this.translateFromProvider(output, params, input, origin_id, logger)

    return [messages, input, metadata]
  }

  async translateToProvider(ellCall: EllCallParams): Promise<OpenAI.Chat.Completions.ChatCompletionCreateParams> {
    const canStream = getCanStream(ellCall)
    if (!canStream && ellCall.apiParams.stream) {
      // TODO. Should we let openai handle this?
      // maybe the user knows better than us (if openai adds support for these arguments)
      throw new Error(
        `Received stream:true api params, but may not be supported for ${ellCall.model} with ${ellCall.tools ? 'tools' : ''} ${ellCall.apiParams.response_format ? 'response_format' : ''}`
      )
    }
    if (canStream) {
      return await mapToStreamingParams(ellCall)
    } else {
      return await mapToNonStreamingParams(ellCall)
    }
  }

  async translateFromProvider(
    provider_response: any,
    ell_call: EllCallParams,
    provider_call_params: Record<string, any>,
    origin_id?: string,
    logger?: (content: string, options: { isRefusal: boolean }) => void
  ): Promise<[Message[], Metadata]> {
    const metadata: Metadata = {}
    const messages: Message[] = []
    const did_stream = provider_call_params.stream || false

    if (did_stream) {
      const message_streams: Record<number, any[]> = {}
      let role: string | undefined

      for await (const chunk of provider_response) {
        Object.assign(metadata, omit(chunk, ['choices']))

        for (const chat_compl_chunk of chunk.choices) {
          const index = chat_compl_chunk.index
          message_streams[index] = message_streams[index] || []
          message_streams[index].push(chat_compl_chunk)
          const delta = chat_compl_chunk.delta
          role = role || delta.role

          if (index === 0 && logger) {
            logger(delta.content || '', { isRefusal: delta.refusal || false })
          }
        }
      }

      for (const [_, message_stream] of Object.entries(message_streams).sort((a, b) => Number(a[0]) - Number(b[0]))) {
        const text = message_stream.map((choice) => choice.delta.content || '').join('')
        messages.push(
          new Message(role!, [
            new ContentBlock({
              text: text,
              // new _lstr(text, origin_id)
            }),
          ])
        )
      }
    } else {
      Object.assign(metadata, omit(provider_response, ['choices']))

      for (const oai_choice of provider_response.choices) {
        const role = oai_choice.message.role
        const content_blocks: ContentBlock[] = []
        const message = oai_choice.message

        if (message.refusal) {
          throw new Error(message.refusal)
        }

        if ('parsed' in message) {
          if (message.parsed) {
            content_blocks.push(new ContentBlock({ parsed: message.parsed }))
            if (logger) logger(JSON.stringify(message.parsed), { isRefusal: false })
          }
        } else {
          if (message.content) {
            content_blocks.push(
              new ContentBlock({
                text: message.content,
                // new _lstr(message.content, origin_id)
              })
            )
            if (logger) logger(message.content, { isRefusal: false })
          }

          if (message.tool_calls) {
            for (const tool_call of message.tool_calls) {
              const matching_tool = ell_call.tools?.find((tool) => tool.__ell_tool_name__ === tool_call.function.name)
              if (!matching_tool) {
                throw new Error('Model called tool not found in provided toolset.')
              }
              content_blocks.push(
                new ContentBlock({
                  tool_call: new ToolCall(
                    matching_tool,
                    tool_call.id,
                    //new _lstr(tool_call.id, origin_id),
                    JSON.parse(tool_call.function.arguments)
                  ),
                })
              )
              if (logger) logger(JSON.stringify(tool_call), { isRefusal: false })
            }
          }
        }

        messages.push(new Message(role, content_blocks))
      }
    }

    return [messages, metadata]
  }
}

const OpenAIProvider = new _OpenAIProvider()

registerProvider(OpenAIProvider, OpenAI)

export const contentBlockToOpenAIFormat = async (
  contentBlock: ContentBlock
): Promise<OpenAI.Chat.Completions.ChatCompletionContentPart> => {
  if (contentBlock.image) {
    const base64Image = await contentBlock.image.serialize()
    return {
      type: 'image_url',
      image_url: { url: base64Image, detail: contentBlock.image.detail },
    } as OpenAI.Chat.Completions.ChatCompletionContentPartImage
  } else if (contentBlock.text) {
    return {
      type: 'text',
      text: contentBlock.text,
    }
  } else if (contentBlock.parsed) {
    return {
      type: 'text',
      text: JSON.stringify(contentBlock.parsed),
    }
  } else {
    throw new Error(`Unsupported content block type for openai: ${contentBlock}`)
  }
}

export const messageToOpenAIFormat = async (
  message: Message
): Promise<OpenAI.Chat.Completions.ChatCompletionMessageParam> => {
  if (message.toolCalls?.length) {
    try {
      if (message.role !== 'assistant') {
        throw new Error('Tool calls must be from the assistant.')
      }
      return {
        role: 'assistant',
        tool_calls: message.toolCalls.map((toolCall) => ({
          id: toolCall.tool_call_id!,
          type: 'function',
          function: {
            name: toolCall.tool.name,
            arguments: JSON.stringify(toolCall.params),
          },
        })),
      }
    } catch (e) {
      console.error(`Error serializing tool calls: ${e}. Did you fully type your @ell.tool decorated functions?`)
      throw e
    }
  }
  if (message.toolResults?.length) {
    return {
      role: 'tool' as const,
      tool_call_id: message.toolResults[0].tool_call_id,
      // @ts-ignore may be undefined
      content: message.toolResults[0].result[0].text!,
    }
  }

  // @ts-ignore
  const openaiMessage:
    | OpenAI.Chat.Completions.ChatCompletionSystemMessageParam
    | OpenAI.Chat.Completions.ChatCompletionUserMessageParam
    | OpenAI.Chat.Completions.ChatCompletionAssistantMessageParam = {
    role: message.role as 'user' | 'assistant' | 'system',
    content: await Promise.all(message.content.map((c) => contentBlockToOpenAIFormat(c))),
  }

  return openaiMessage
}
